### PR TITLE
Refactor createTemplateService and add a few automated tests

### PR DESCRIPTION
**Pull request by anarute**
This PR should be read as a proposal for a bigger architectural refactor.

Since I need to work on a big feature, I really want us to cover the project with automated tests but to achieve this goal we need to do some refactoring.

I used the `createTemplateService` as the first gnea pig since I have worked a bit with it and it is a bit independent.

I'm still not 100% happy with the name `TemplateService` and even if it should be inside the services folder, but for now maybe it's enough.

My idea is to slowly try to apply [DDD](https://en.wikipedia.org/wiki/Domain-driven_design) in the project, but I want to be conservative for now to avoid breaking things by mistake.

In summary in this PR:
- Remove the test folder and all the existing tests (they didn't seem to be testing anything really)
- Refactor `createTemplatesService` to call a new class `TemplateService` that tries to break down a bit how we are parsing the template creation request
- Add 2 initial tests to test the `TemplateService` class
- Introduce `namespace` to the projects (I'm still learning so I know it can be better hehe )

### CODE REVIEW

**Comment by jaragunde** [comment]
Hi Ana, thanks for this hard work. I need to set aside some time to go through your proposal, but I have one comment.

Could the new autoload setup be used to simplify the usage of require/include we do in the codebase? If you take a look, you will find that any user-facing PHP pages define this constant pointing to the root directory:
```php
define('PHPREPORT_ROOT', __DIR__ . '/../../');
```
Then, every include uses it like:
```php
include_once(PHPREPORT_ROOT . '/model/vo/TemplateVO.php');
```

So, with autoload and namespaces, could we stop using `PHPREPORT_ROOT`? Notice I'm a bit lost about modern PHP practices, too :-)

**Comment by anarute** [comment]
> Hi Ana, thanks for this hard work. I need to set aside some time to go through your proposal, but I have one comment.
> 
> Could the new autoload setup be used to simplify the usage of require/include we do in the codebase? If you take a look, you will find that any user-facing PHP pages define this constant pointing to the root directory:
> 
> ```
> define('PHPREPORT_ROOT', __DIR__ . '/../../');
> ```
> 
> Then, every include uses it like:
> 
> ```
> include_once(PHPREPORT_ROOT . '/model/vo/TemplateVO.php');
> ```
> 
> So, with autoload and namespaces, could we stop using `PHPREPORT_ROOT`? Notice I'm a bit lost about modern PHP practices, too :-)

That's a good point, I'll create an autoload file and use it instead of the one generated by composer so we don't need to use compose to deploy to production yet, only to run the automated tests locally. This will be enough for us to clean up these include lines :) I'll work on it next week (I'll be off for a few days)

**Comment by anarute** [comment]
And yes we'll be able to stop using `PHPREPORT_ROOT` 🎉

**Review by jaragunde — COMMENTED** [review]
I agree with the general direction in this PR, and I'm eager to see how we can share more code between different services, add more tests, etc.

I just have a few comments below:

**Comment by jaragunde** [comment]
> That's a good point, I'll create an autoload file and use it instead of the one generated by composer so we don't need to use compose to deploy to production yet, only to run the automated tests locally. This will be enough for us to clean up these include lines :) I'll work on it next week (I'll be off for a few days)

I finally understood what you meant when I read the last commit (I hadn't noticed it when I wrote the review above).

If I understand correctly, it's a temporary solution to prevent composer from being compulsory, right? If that's the case, I would even prefer to start using composer and the improved autoload at the same time and skip the temporary situation.

File: web/services/TemplateService.php [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+if (!defined('PHPREPORT_ROOT')) define('PHPREPORT_ROOT', __DIR__ . '/../../');
+
+include_once(PHPREPORT_ROOT . '/web/services/WebServicesFunctions.php');
+include_once(PHPREPORT_ROOT . '/model/facade/TemplatesFacade.php');
+
+class TemplateService
+{
+    private \TemplateVO $templateVO;
+    private \LoginManager $loginManager;
```
[the remark this answers is not in the document]
Comment (anarute) [reply web/services/TemplateService.php:31]: Yes, will do! I'll try to break this down into smaller commits

**Comment by anarute** [comment]
> > That's a good point, I'll create an autoload file and use it instead of the one generated by composer so we don't need to use compose to deploy to production yet, only to run the automated tests locally. This will be enough for us to clean up these include lines :) I'll work on it next week (I'll be off for a few days)
> 
> I finally understood what you meant when I read the last commit (I hadn't noticed it when I wrote the review above).
> 
> If I understand correctly, it's a temporary solution to prevent composer from being compulsory, right? If that's the case, I would even prefer to start using composer and the improved autoload at the same time and skip the temporary situation.

Yes, it would be an intermediate step, ok I'll remove this commit and keep composer :)

File: composer.lock [hunk 2]
```diff
@@ -0,0 +1,2121 @@
+{
+    "_readme": [
+        "This file locks the dependencies of your project to a known state",
+        "Read more about it at https://getcomposer.org/doc/01-basic-usage.md#installing-dependencies",
+        "This file is @generated automatically"
+    ],
```
[the remark this answers is not in the document]
Comment (anarute) [reply composer.lock:6]: Yes :/ they are all sub dependencies of `phpunit/phpunit`.
Note that for production there are no dependencies needed, only for dev (under `packages-dev` line 9)

File: web/services/TemplateService.php [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+    public function __construct(
+        \TemplateVO $templateVO,
```
[the remark this answers is not in the document]
Comment (anarute) [reply web/services/TemplateService.php:34]: This will create an object every time, but now it is created outside the class. The idea is to use dependency injection to make the class more independent and easier to isolate and test. 

But for TemplateVo you are probably right that I don't need to do this since it's very isolated already, I'll change this and pass only the LoginManager.

**Comment by anarute** [comment]
@jaragunde I broke down the commits trying to make the change in steps for a better review. Can you review the whole PR again please?

Sorry for making it confusing the first time 😅 

About the usage of `PHPREPORT_ROOT`, we'll be able to replace all the requires and includes by `namespace` and `use` and do a unique include ` require PHPREPORT_ROOT . "vendor/autoload.php";` in some strategic files, but this would require to modify almost every file in the repository, that's why I haven't done yet. I think it's better if we start using composer and make sure this patch works in beta and prod, then I can do for the rest of the project.

File: web/services/TemplateService.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (jaragunde) [reply web/services/TemplateService.php]: I've been reading about dependency injection and, while I can see the advantage of injecting the LoginManager, I'm not sure about the TemplateVO. I wonder what's the usual procedure: inject all classes, decide case-by-case... The main difference for me is that, while LoginManager provides a service, TemplateVO acts more like a basic data type.

> This will create an object every time, but now it is created outside the class. 

I see, now I realize the template create operation only supports creating one template for every operation. But, I wonder how DI could be achieved if the service supported creating several at the same time, like the task create service does.

**Comment by jaragunde** [comment]
> I think it's better if we start using composer and make sure this patch works in beta and prod, then I can do for the rest of the project.

I agree, let's take small steps.

**Comment by jaragunde** [comment]
Note (to self): composer is packaged in Fedora and Debian (probably Ubuntu, I guess), the package name is `composer`. We probably could mention that in the instructions, like we mention package names for other dependencies. I can do that later.

On the other hand, we might just distribute the phpreport releases with the autoload already generated and dependencies included and mention composer only as a development dependency. Let's figure this out later, when we discuss future releases... Maybe it's not even worth to keep doing them!

File: docs/developer/devel-setup.rst [hunk 4]
```diff
@@ -62,3 +62,14 @@ password ``admin`` to try a user with full privileges, or ``user`` with
 password ``user`` to try a login with restricted permissions. You will also be
 able to create users with any permission level once you are logged in as
 ``admin``.
+
+Step 6: Running automated tests
+===========================
+
```
[the remark this answers is not in the document]
Comment (anarute) [reply docs/developer/devel-setup.rst:68]: oh my bad, I didn't notice it had to match the amount of characters of the title

[the remark this answers is not in the document]
Comment (jaragunde) [reply docs/developer/devel-setup.rst:68]: I didn't know, either! I thought it was enough with three. But this is not actually Markdown syntax so :shrug:

**Comment by jaragunde** [comment]
> On the other hand, we might just distribute the phpreport releases with the autoload already generated and dependencies included and mention composer only as a development dependency. Let's figure this out later, when we discuss future releases... Maybe it's not even worth to keep doing them!

I did one last release without the composer dependency, in the old style. I think that, in the future, it's not worth to keep doing these releases, they are a lot of manual, error-prone work. I've opened #511 to discuss that.
